pkg/controllers/metrics/state/node.go

```diff
@@ -0,0 +1,262 @@
+			Help:      "Node total pod limits are the resources specified by non-DaemonSet pod limits. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.",
+		},
+		nodeLabelNames(),
+	)
+
+	daemonRequestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_daemon_limits",
+			Help:      "Node total daemon requests are the resource requested by DaemonSet pods bound to nodes. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.",
+		},
+		nodeLabelNames(),
+	)
+
+	daemonLimitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_daemon_requests",
+			Help:      "Node total daemon limits are the resources specified by DaemonSet pod limits. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.",
+		},
+		nodeLabelNames(),
+	)
+
+	overheadGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "system_overhead",
+			Help:      "Node system daemon overhead are the resources reserved for system overhead, the difference between the node's capacity and allocatable values are reported by the status. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.",
+		},
+		nodeLabelNames(),
+	)
+
+	wellKnownLabels = getWellKnownLabels()
+)
+
+func newNodeGaugeVec(name, help string) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      name,
+			Help:      help,
+		},
+		nodeLabelNames(),
+	)
+}
+
+func nodeLabelNames() []string {
+	labels := []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodePhase, // NOTE: deprecated
+	}
+
+	for _, l := range wellKnownLabels {
+		labels = append(labels, l)
+	}
+
+	return labels
+}
+
+func init() {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		metrics.Registry.MustRegister(gauge)
+	}
+}
+
+type NodeScraper struct {
+	cluster  *state.Cluster
+	labelMap map[string]map[*prometheus.GaugeVec][]prometheus.Labels
+}
+
+func NewNodeScraper(cluster *state.Cluster) *NodeScraper {
+	return &NodeScraper{
+		cluster:  cluster,
+		labelMap: make(map[string]map[*prometheus.GaugeVec][]prometheus.Labels),
+	}
+}
+
+func (ns *NodeScraper) Scrape(ctx context.Context) {
+	nodes := sets.NewString()
+	ns.cluster.ForEachNode(func(n *state.Node) bool {
+		if _, ok := ns.labelMap[n.Node.Name]; !ok {
+			ns.labelMap[n.Node.Name] = make(map[*prometheus.GaugeVec][]prometheus.Labels)
+		}
+		nodes.Insert(n.Node.Name)
```

Comment:
It strikes me that if the cluster's interface was GetNodes() instead of a callback approach ForEachNode(), you wouldn't need to maintain this list separately.